### PR TITLE
Canonical Projects

### PR DESCRIPTION
**Pull request by 0xKurt**
### Description 

This PR speed up the builder project and reduces graph calls by 3x.
It is meant to allow you to create a linked project behind the scenes when you apply to a round on a different chain.
This PR should be tested only for QF 

#### Allo V1

- [x] Create Project
- [x] Apply To Round (same chain)
- [x] Apply To Round (different chain)
- [x] View Application
- [x] Vew Project / Stats / Rounds
- [x] Edit a Project

####  Allo V2

- [x] Create Project
- [x] Apply To Round (same chain)
- [x] Apply To Round (different chain)
- [x] Apply To Round (canonical project in diff chain but linked project exists)
- [x] View Application 
- [x] Vew Project / Stats / Rounds
- [x] Edit a Project

#### Open rounds to apply:
[v1 round](https://builder-staging-git-canonical-projects-grants-stack.vercel.app/#/chains/11155111/rounds/0x0a651f9aba8be2280557013d0bcdfeeb44294b19) (staging link)
[v2 round](https://builder-staging-git-canonical-projects-grants-stack.vercel.app/#/chains/11155111/rounds/20) (staging link)

Fixes: https://github.com/gitcoinco/grants-stack/issues/2857
Fixes: https://github.com/gitcoinco/grants-stack/issues/2928
Fixes https://github.com/gitcoinco/grants-stack/issues/2911

Video part 1: https://drive.google.com/file/d/1CC-6xqcAQXH6Gv4yQLAnex2wZSEGwWPr/view?usp=drive_link
Video part 2: https://drive.google.com/file/d/1lb6515qdGdlNlrhrRoFHbrb4VHmdSIa_/view?usp=drive_link

### CODE REVIEW

File: packages/data-layer/src/data-layer.test.ts [hunk 1]
```diff
@@ -605,16 +611,20 @@ describe("projects retrieval", () => {
...
-      chainIds: [5],
+      chainIds: [11155111],
       first: 10,
       alloVersion: "allo-v2",
     });
 
-    if (data?.projects) expect(data.projects.length).toBeGreaterThan(0);
+    console.log(data);
+
+    // if (data?.projects) expect(data.projects.length).toBeGreaterThan(0);
```
Comment (codenamejason) [inline packages/data-layer/src/data-layer.test.ts:627]: todo

**Review by boudra — CHANGES_REQUESTED** [review]
i see some console logs and todos, please do self review first and let me know

File: packages/builder/src/reducers/roundApplication.ts [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
   APPLICATION_DATA_LOADED,
-  RoundApplicationActions,
   ROUND_APPLICATION_ERROR,
   ROUND_APPLICATION_ERROR_RESET,
   ROUND_APPLICATION_FOUND,
   ROUND_APPLICATION_LOADED,
   ROUND_APPLICATION_LOADING,
   ROUND_APPLICATION_NOT_FOUND,
   ROUND_APPLICATION_RESET,
+  RoundApplicationActions,
```
Comment (codenamejason) [inline packages/builder/src/reducers/roundApplication.ts:10]: This naming seems odd here 🤔

**Review by codenamejason — APPROVED** [review]
Nice work everyone 🌱

**Review by michellema1208 — APPROVED** [review]
@0xKurt / @codenamejason / @thelostone-mc -- TY so much for adding the v1 and v2 staging links! I used both of those and was able to successfully apply cross-chain. 

v1: used a PGN testnet project to apply to Sepolia round. Successful tx: https://sepolia.etherscan.io/tx/0x22d5bad744dd1e6af59acda82405036e0bb31580aad0a06f140e30541cde1d77

v2: used Polygon Mumbai project to apply to Sepolia round. Successful tx: https://sepolia.etherscan.io/tx/0x7590f7d88136a0247071a9d97e671ccc15c11201b4795a5d2d9684a5ec069d2a

File: packages/data-layer/src/data-layer.ts [hunk 3]
```diff
@@ -411,6 +356,58 @@ export class DataLayer {
...
+   * Returns a single application as identified by its id, round name and chain name
+   * @param projectId
+   */
+  async getApplicationsByRoundIdAndProjectIds({
+    chainId,
+    roundId,
+    projectIds,
+  }: {
+    chainId: number;
+    roundId: Lowercase<Address>;
```
Comment (boudra) [inline packages/data-layer/src/data-layer.ts:369]: round ID is not always and address, string would be more appropriate here

File: packages/data-layer/src/data-layer.ts [hunk 4]
```diff
@@ -411,6 +356,58 @@ export class DataLayer {
...
+    projectIds: string[];
+  }): Promise<ProjectApplication[] | undefined> {
```
Comment (boudra) [inline packages/data-layer/src/data-layer.ts:371]: this function never returns undefined

File: packages/data-layer/src/data-layer.ts [hunk 5]
```diff
@@ -411,6 +356,58 @@ export class DataLayer {
...
+    roundId: string;
+    metadataCid: string;
+  }): Promise<ApplicationStatus | undefined> {
+    const requestVariables = {
+      chainId,
+      roundId,
+      metadataCid,
+    };
+
+    const response: { applications: any } = await request(
```
Comment (boudra) [inline packages/data-layer/src/data-layer.ts:402]: any are not allowed

File: packages/data-layer/src/data-layer.ts [hunk 6]
```diff
@@ -411,6 +356,58 @@ export class DataLayer {
...
+    const response: { applications: ProjectApplication[] } = await request(
+      this.gsIndexerEndpoint,
+      getApplicationsByRoundIdAndProjectIds,
+      requestVariables,
+    );
+
+    return response.applications ?? [];
+  }
+
+  async getApplicationStatusByRoundIdAndCID({
```
Comment (boudra) [inline packages/data-layer/src/data-layer.ts:387]: this function doesn't seem to be used

File: packages/data-layer/src/data.types.ts [hunk 7]
```diff
@@ -238,6 +249,10 @@ export type V2Round = {
   projectId: string;
   strategyAddress: string;
   strategyName: string;
+  project?: {
```
Comment (boudra) [inline packages/data-layer/src/data.types.ts:252]: this should be it's own type like ProjectApplicationWithProject

File: packages/builder/src/components/rounds/ViewApplication.tsx [hunk 8]
```diff
@@ -72,13 +67,13 @@ function ViewApplication() {
       round,
       applicationState,
       applicationStatus,
-      applicationError,
+      // applicationError,
```
Comment (boudra) [inline packages/builder/src/components/rounds/ViewApplication.tsx:70]: commented out code shouldn't be in production

**Pull request merged**